---
title: "Analysis of cell type from single cell RNA-seq data"
author: "Diego Diez"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(dplyr)
library(celltype)
library(SingleCellExperiment)
```

We use a `SingleCellExperiment` object.

```{r}
sce1
```

We use a database of gene markers from the Immgen project.

```{r}
head(markers)
```

We use a dictionary of gene expression for the gene markers in different cell types.

```{r}
head(immgen.db)
```

Obtain a matrix of cell type correlations.

```{r}
celltype <- predict_celltype(sce1)
celltype[1:5, 1:3]
```

We choose the cell type based on maximum correlation.

```{r}
celltype <- choose_celltype(celltype)
head(celltype)
```

For Immgen dictionary we can focus on the top level cell hierarchy.

```{r}
celltype_simple <- fix_immgen_celltype(celltype)

celltype <- celltype %>%mutate(celltype_simple = celltype_simple)

head(celltype)
```

We can use alternative dictionaries.

```{r}
celltype <- predict_celltype(sce1, name = "immnav")
celltype[1:5, 1:2]
```

And again choose the cell type based on maximum correlation.

```{r}
celltype <- choose_celltype(celltype)
head(celltype)
```
